### PR TITLE
Making CI for testing generated proto correctness

### PR DESCRIPTION
**Pull request by xBalbinus**
Closes: [#1256](https://github.com/osmosis-labs/osmosis/issues/1256)

## What is the purpose of the change

This CI will help us catch whenever make proto-all wasn't ran via a job that does make proto-all,and returns an error if there is a diff betrween the post make proto-all output, and any .proto or .pb.go file, as well as tool-chain related errors.

## Brief Changelog

Added job in check-generated.yml that fails if make proto-all wasn't done before PR

## Testing and Verifying

*(Please pick one of the following options)*

This change added tests and can be verified as follows:

*(example:)*
- Demo CI run: https://github.com/xBalbinus/osmosis/actions/runs/2583435511

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / **no**)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / **no**)
  - How is the feature or change documented? (**not applicable**   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Review by mattverse — CHANGES_REQUESTED** [review]
Can we make it so that we do not check protobuf generated code for `go.mod` and `go.sum`?

File: .github/workflows/check-generated.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          go-version: '1.18'
+
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 1  # we need a .git directory to run git diff
+
+      - name: "Check protobuf generated code"
+        run: |
+          set -euo pipefail
```
Comment (alexanderbez) [inline .github/workflows/check-generated.yml:30]: We should just put this all in a script under `contrib/`.

**Review by alexanderbez — CHANGES_REQUESTED** [review]

File: .github/workflows/check-generated.yml (hunk 1 above)
Comment (xBalbinus) [reply .github/workflows/check-generated.yml:30]: sgtm! doing so.

**Comment by xBalbinus** [comment]
Looking into the go.mod issue!

File: .github/workflows/check-generated.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 1  # we need a .git directory to run git diff
+          
+      - name: "Check protobuf generated code"
+        run: contrib/scripts/check-generated.sh
+        shell: bash
+        with:
+          password: ${{ secrets.ADD_TO_PROJECT_PAT }}
```
[the remark this answers is not in the document]
Comment (xBalbinus) [reply .github/workflows/check-generated.yml:32]: This most likely isn't - sorry I'm just testing out some stuff to implement the feedback above

**Comment by mattverse** [comment]
Oh I was saying that this "shouldn't" include `go.mod` and `go.sum`, since we're maintaining them separately?

**Review by alexanderbez — COMMENTED** [review]
🚀

**Comment by xBalbinus** [comment]
All feedback implemented, ready for review / lmk if I missed any :)

Thanks for the great advice folks.

Working CI run for ref: https://github.com/xBalbinus/osmosis/runs/7141298212?check_suite_focus=true

**Comment by mattverse** [comment]
Seems like the CI is breaking for this PR? Is that an expected behaviour resulting from this PR not being merged yet?

**Comment by xBalbinus** [comment]
Unexpected behavior, I detailed this a bit in slack if you wouldn’t mind
taking a look :)

File: .github/workflows/check-generated.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Note that we run these checks regardless whether the input files have
+# changed, because generated code can change in response to toolchain updates
+# even if no files in the repository are modified.
+name: Check generated code
+on:
+  workflow_dispatch:
+  pull_request:
+    branches:
+      - main
```
Comment (ValarDragon) [inline .github/workflows/check-generated.yml:11]: Can we do `*`, otherwise LGTM

Comment (xBalbinus) [reply .github/workflows/check-generated.yml]: Done and dusted 👍

**Review by mattverse — APPROVED** [review]
let's go 🔥

**Review by alexanderbez — APPROVED** [review]
utACK 🎉

**Pull request merged**